sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -171,10 +171,13 @@ abstract class Optimizer(sessionCatalog: SessionCatalog)
     // "Extract PythonUDF From JoinCondition".
     Batch("Check Cartesian Products", Once,
       CheckCartesianProducts) :+
-    Batch("RewriteSubquery", Once,
+    Batch("Rewrite Subquery", Once,
```

Comment:
I do not have a good answer for this PR. Ideally, we should run the whole batch `operatorOptimizationBatch`. However, running the rules could be very time consuming. I would suggest to add a new parameter for introducing the time bound limit for each batch.

cc @maryannxue  WDYT?

Reply:
@gatorsmile Do you think its a good time to revisit Natt's PR to convert subquery expressions to Joins early in the optimization process ?  Perhaps then we can take advantage of all the subsequent rules firing after the subquery rewrite ?

Reply:
@gatorsmile I think @dilipbiswal's suggestion is the right way to go.

Reply:
Sure. That sounds also good to me. @dilipbiswal Could you take the PR https://github.com/apache/spark/pull/17520 over?

Reply:
@gatorsmile Sure Sean.. Let me give it a try.